### PR TITLE
Use path length to skip scanning subcollections

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This is a new version of https://github.com/firebase/firebase-android-sdk/pull/3173 that uses a "path segment count" to also skip scanning subcollections. The main benefit is that it stores less data.

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java [hunk 1]
```diff
@@ -138,88 +135,73 @@ public MutableDocument get(DocumentKey documentKey) {
...
-    ResourcePath prefix = query.getPath();
-    int immediateChildrenPathLength = prefix.length() + 1;
+    StringBuilder sql =
+        new StringBuilder(
+            "SELECT contents, read_time_seconds, read_time_nanos "
+                + "FROM remote_documents WHERE path >= ? AND path < ? AND path_length = ?");
 
-    String prefixPath = EncodedPath.encode(prefix);
-    String prefixSuccessorPath = EncodedPath.prefixSuccessor(prefixPath);
+    Object[] bindVars = new Object[3 + (FieldIndex.IndexOffset.NONE.equals(offset) ? 0 : 6)];
```
Comment (wu-hui) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java:143]: Give `FieldIndex.IndexOffset.NONE.equals(offset)` a name?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java [hunk 2]
```diff
@@ -607,6 +629,35 @@ private void rewriteCanonicalIds() {
             });
   }
 
+  /** Fill the remote_document's path_length column. */
```
Comment (wu-hui) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:632]: Ultranit: `Fills` or `Populates`.

**Review by wu-hui — APPROVED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java:143]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java (hunk 2 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:632]: Done

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks for the review

**Pull request merged**